src/poprox_recommender/default.py

```diff
@@ -34,16 +42,25 @@ def select_articles(
     if pipeline is None:
         pipeline = fallback_pipeline(num_slots)
 
-    inputs = {
-        "candidate": candidate_articles,
-        "clicked": clicked_articles,
-        "profile": interest_profile,
-    }
+    rank = pipeline.node("rank")
+    topk = pipeline.node("rank-topk", missing="none")
+    if topk is None:
+        wanted = (rank,)
+    else:
+        wanted = (rank, topk)
+
+    result = pipeline.run(*wanted, candidate=candidate_articles, clicked=clicked_articles, profile=interest_profile)
 
-    return pipeline(inputs)
+    if topk is None:
+        assert isinstance(result, ArticleSet)
+        return Recommendations(result.articles, initial=result)
+    else:
+        assert isinstance(result, tuple)
+        reranked, initial = result
```

Comment:
Relying on tuple order here seems a potential smell—if we're going to return a data structure, it should probably be at least a `namedtuple` or `dataclass`

Reply:
Also not wild about returning different types in different circumstances

Reply:
<sidebar> This is another place where I really wish Python's type system had named types. With named types, we could do fun things like let the input be a `TypedDict` of nodes and the output is a `TypedDict` of return data.

Right now `run` has several overloads so that its return type is statically inferred based on the node object types, when those are available (up to 5 requested node outputs).